File: 0/1/1/011.py
# Copyright (C) 2013 Jorge Aparicio

import numpy as np

dirs = [(1, 0), (0, 1), (1, 1), (1, -1)]
size = 20
stride = 4

with open('011.in') as f:
    arr = np.asarray(list(map(int, f.read().split())), dtype=np.int_)
    arr = arr.reshape(size, size)

ans = 0
for dir_ in dirs:
    for i in range(size):
        if not 0 <= i + (stride - 1) * dir_[0] < size:
            continue
        for j in range(size):
            if not 0 <= j + (stride - 1) * dir_[1] < size:
                continue
            p = arr[i, j]
            for k in range(1, stride):
                p *= arr[i + k * dir_[0], j + k * dir_[1]]

            if p > ans:
                ans = p

print(ans)

File: 0/1/1/011.rs
// Copyright (C) 2014 Jorge Aparicio

use std::io::fs::File;
use std::path::posix::Path;

static stride: uint = 4;


fn main() {
  let path = Path::new("011.in");
  let mut file = File::open(&path);
  let contents = file.read_to_str();

  let grid =
    contents.
    lines().
    map(|line| line.words().filter_map(from_str::<int>).to_owned_vec()).
    to_owned_vec();

  let (nr, nc) = (grid.len(), grid[0].len());
  let directions = ~[(0, 1), (1, 0), (1, 1), (-1, 1)];

  let mut max= 0;
  for &(y, x) in directions.iter() {
    for r in range(0, nr) {
      if r + stride * y > nr { continue }

      for c in range(0, nc) {
        if c + stride * x > nc { continue }

        let mut tmp = grid[r][c];
        for i in range(1, stride) {
          tmp *= grid[r + i * y][c + i * x];
        }

        if tmp > max {
          max = tmp;
        }
      }
    }
  }

  println!("{}", max);
}

